File: mobile/test/scheme_signup_test.dart
import 'package:flutter_test/flutter_test.dart';
import 'package:health_info_app/pharmacy.dart';

void main() {
  test('the scheme half rides nested; the admin half sits beside it', () {
    final body = schemeSignUpBody(
      tenantId: 7,
      name: '  AXA Mansard ',
      code: ' AXA ',
      coveragePercent: '80.00',
      preauthThreshold: '5000',
      autoSubmitClaims: true,
      adminPhone: ' 08031234567 ',
      adminName: ' Ada ',
      adminPassword: 's3curepass99',
    );
    expect(body['tenant'], 7);
    expect(body['scheme'], {
      'name': 'AXA Mansard',
      'code': 'AXA',
      'contact': '',
      'email': '',
      'coverage_percent': '80.00',
      'preauth_threshold': '5000',
      'auto_submit_claims': true,
    });
    expect(body['admin_phone'], '08031234567');
    expect(body['admin_name'], 'Ada');
    expect(body['admin_password'], 's3curepass99');
  });

  test('a threshold left blank is 0 — never ask first, not no cover', () {
    final body = schemeSignUpBody(
      tenantId: 1,
      name: 'Hygeia',
      coveragePercent: '  ',
      preauthThreshold: '   ',
      adminPhone: '08031234568',
      adminPassword: 's3curepass99',
    );
    final scheme = body['scheme'] as Map<String, dynamic>;
    expect(scheme['preauth_threshold'], '0');
    expect(scheme['coverage_percent'], '100');
  });

  test('a password is sent as typed — trimming one changes it', () {
    final body = schemeSignUpBody(
      tenantId: 1,
      name: 'Reliance',
      adminPhone: '08031234569',
      adminPassword: ' spaced pass 99 ',
    );
    expect(body['admin_password'], ' spaced pass 99 ');
  });

  Map<String, dynamic> ready({
    int? tenantId = 1,
    String name = 'Hygeia',
    String adminPhone = '08031234567',
    String adminPassword = 's3curepass99',
  }) =>
      schemeSignUpBody(
        tenantId: tenantId,
        name: name,
        adminPhone: adminPhone,
        adminPassword: adminPassword,
      );

  test('a body with both halves filled in is ready to send', () {
    expect(schemeSignUpProblem(ready()), isNull);
  });

  test('each half the API would refuse is named before the round trip', () {
    expect(schemeSignUpProblem(ready(tenantId: null)), contains('organization'));
    expect(schemeSignUpProblem(ready(name: '  ')), 'Name the scheme.');
    expect(schemeSignUpProblem(ready(adminPhone: ' ')), contains('phone number'));
    expect(
        schemeSignUpProblem(ready(adminPassword: 'short')), contains('8 characters'));
  });

  test('an insurer filed without a seat is only checked on the scheme half', () {
    final bare = ready(adminPhone: '', adminPassword: '');
    expect(schemeSignUpProblem(bare, withAdmin: false), isNull);
    expect(schemeSignUpProblem(bare), contains('phone number'));
    // The organization and the name still have to hold up: that row is a
    // scheme the counter prices sales off.
    expect(schemeSignUpProblem(ready(tenantId: null), withAdmin: false),
        contains('organization'));
  });
}

File: web/schemesignup.test.js
/* Self-check for the scheme sign-up screen in app.js. No DOM beyond FormData,
 * which node has: the pieces worth checking are plain over plain values.
 * Run with: node schemesignup.test.js
 *
 * ponytail: source sliced out of app.js, same as scheme.test.js — the app is
 * loaded with <script> tags and has no build step to import from. */
'use strict';
const assert = require('assert');
const { readFileSync } = require('fs');

const src = readFileSync(`${__dirname}/app.js`, 'utf8');

// The screen is only reachable if the route and the way in are both wired.
assert.ok(/scheme-register\$\/, viewSchemeRegister/.test(src),
  'scheme-register route missing');
assert.ok(/res\.signUp && ME\?\.role === 'super_admin'/.test(src),
  'no way in from the list screens');

const from = src.indexOf('function schemeSignUpBody(');
const to = src.indexOf('async function viewProfile()');
assert.ok(from > 0 && to > from, 'sign-up helpers not found in app.js');
const [schemeSignUpBody, schemeSignUpProblem, flattenSchemeErrors] =
  new Function(`${src.slice(from, to)};
    return [schemeSignUpBody, schemeSignUpProblem, flattenSchemeErrors];`)();

const form = (fields) => {
  const fd = new FormData();
  for (const [k, v] of Object.entries(fields)) fd.set(k, v);
  return fd;
};

const filled = {
  tenant: '7', name: '  AXA Mansard ', code: ' AXA ',
  coverage_percent: '80.00', preauth_threshold: '5000',
  auto_submit_claims: 'on',
  admin_phone: ' 08031234567 ', admin_name: ' Ada ',
  admin_password: 's3curepass99',
};

// The scheme half rides nested (it is the ordinary insurer serializer); the
// admin half sits beside it.
const body = schemeSignUpBody(form(filled));
assert.strictEqual(body.tenant, '7');
assert.deepStrictEqual(body.scheme, {
  name: 'AXA Mansard', code: 'AXA', contact: '', email: '',
  coverage_percent: '80.00', preauth_threshold: '5000',
  auto_submit_claims: true,
});
assert.strictEqual(body.admin_phone, '08031234567');
assert.strictEqual(body.admin_name, 'Ada');

// Left alone, the two decimals are the scheme's defaults — a blank threshold
// is 0 ("never ask first"), not no cover.
const bare = schemeSignUpBody(form({
  tenant: '1', name: 'Hygeia', coverage_percent: '  ', preauth_threshold: '  ',
  admin_phone: '08031234568', admin_password: 's3curepass99',
}));
assert.strictEqual(bare.scheme.preauth_threshold, '0');
assert.strictEqual(bare.scheme.coverage_percent, '100');
assert.strictEqual(bare.scheme.auto_submit_claims, false);

// A password is sent as typed. Trimming one changes it.
assert.strictEqual(
  schemeSignUpBody(form({ ...filled, admin_password: ' spaced pass 99 ' })).admin_password,
  ' spaced pass 99 ');

// What the round trip is saved for.
assert.strictEqual(schemeSignUpProblem(body), null);
for (const [field, says] of [
  ['tenant', /organization/],
  ['name', /Name the scheme/],
  ['admin_phone', /phone number/],
]) {
  const short = schemeSignUpBody(form({ ...filled, [field]: '' }));
  assert.match(schemeSignUpProblem(short), says, `${field} passed unchecked`);
}
assert.match(
  schemeSignUpProblem(schemeSignUpBody(form({ ...filled, admin_password: 'short' }))),
  /8 characters/);

// The scheme's own errors arrive nested; the form is flat, so they are lifted
// to sit beside the admin's.
assert.deepStrictEqual(
  flattenSchemeErrors({ scheme: { name: ['Already taken.'] }, admin_phone: ['Taken.'] }),
  { name: ['Already taken.'], admin_phone: ['Taken.'] });

// A non-field error on the nested serializer comes back as a list, not a dict:
// there is no field to lift it onto, so it is left where it is.
const listed = { scheme: ['Invalid data.'] };
assert.deepStrictEqual(flattenSchemeErrors(listed), listed);

// Nothing to lift, and nothing at all, both pass straight through.
assert.deepStrictEqual(flattenSchemeErrors({ tenant: ['Required.'] }), { tenant: ['Required.'] });
assert.strictEqual(flattenSchemeErrors(null), null);

console.log('schemesignup.test.js ok');
